File: example-codes/atm/customer.py
# For simplicity, we are not defining getter and setter functions. The reader can
# assume that all class attributes are private and accessed through their respective
# public getter methods and modified only through their public methods function.


class Customer:
    def __init__(self, name, address, email, phone, status):
        self.__name = name
        self.__address = address
        self.__email = email
        self.__phone = phone
        self.__status = status
        self.__card = Card()
        self.__account = Account

    def make_transaction(self, transaction):
        None

    def get_billing_address(self):
        None


class Card:
    def __init__(self, number, customer_name, expiry, pin):
        self.__card_number = number
        self.__customer_name = customer_name
        self.__card_expiry = expiry
        self.__pin = pin

    def get_billing_address(self):
        None


class Account:
    def __init__(self, account_number):
        self.__account_number = account_number
        self.__total_balance = 0.0
        self.__available_balance = 0.0

    def get_available_balance(self):
        return self.__available_balance


class SavingAccount(Account):
    def __init__(self, withdraw_limit):
        self.__withdraw_limit = withdraw_limit


class CheckingAccount(Account):
    def __init__(self, debit_card_number):
        self.__debit_card_number = debit_card_number

File: example-codes/atm/java/Customer.java
package com.bank;

/**
 * Represents a bank customer.
 */
public class Customer {
    private String name;
    private Constants.Address address;
    private String email;
    private String phone;
    private Constants.CustomerStatus status;
    private Card card;
    private Account account;

    public Customer(String name, Constants.Address address, String email, String phone,
                    Constants.CustomerStatus status, Card card, Account account) {
        this.name = name;
        this.address = address;
        this.email = email;
        this.phone = phone;
        this.status = status;
        this.card = card;
        this.account = account;
    }

    public String getName() {
        return name;
    }

    public void makeTransaction(Transaction transaction) {
        transaction.makeTransaction();
    }
}

/**
 * Represents a customer's card.
 */
class Card {
    private String number;
    private String customerName;
    private String expiry;
    private String pin;

    public Card(String number, String customerName, String expiry, String pin) {
        this.number = number;
        this.customerName = customerName;
        this.expiry = expiry;
        this.pin = pin;
    }
}

/**
 * Represents an account base class.
 */
class Account {
    protected String accountNumber;
    protected double balance;

    public Account(String accountNumber) {
        this.accountNumber = accountNumber;
        this.balance = 0.0;
    }

    public double getBalance() {
        return balance;
    }
}

/**
 * Subclass for Savings account.
 */
class SavingAccount extends Account {
    private double withdrawLimit;

    public SavingAccount(String accountNumber, double withdrawLimit) {
        super(accountNumber);
        this.withdrawLimit = withdrawLimit;
    }
}

/**
 * Subclass for Checking account.
 */
class CheckingAccount extends Account {
    private String debitCardNumber;

    public CheckingAccount(String accountNumber, String debitCardNumber) {
        super(accountNumber);
        this.debitCardNumber = debitCardNumber;
    }
}
